File: ios/Classes/VolcEngineSpeechContent.swift
//
//  VolcEngineSpeechContent.swift
//  volc_engine_asr
//
//  Created by cplin on 2024/8/6.
//

import Foundation

enum VolcEngineSpeechContentType: Int, Codable {
    case volume = 0
    case text = 1
    case recordStatus = 2
}

struct VolcEngineSpeechContent: Codable {
    let type: VolcEngineSpeechContentType
    let volume: Double
    let text: String
    let duration: Int
    let isRecording: Bool
    
    init(type: VolcEngineSpeechContentType, volume: Double = 0.0, text: String = "", duration: Int = 0, isRecording: Bool) {
        self.type = type
        self.volume = volume
        self.text = text
        self.duration = duration
        self.isRecording = isRecording
    }
    
    func toJson() -> String? {
        let encoder = JSONEncoder()
        encoder.outputFormatting = .prettyPrinted
        if let jsonData = try? encoder.encode(self) {
            return String(data: jsonData, encoding: .utf8)
        }
        return nil
    }
    
    static func volume(volume: Double) -> VolcEngineSpeechContent {
        return VolcEngineSpeechContent(type: .volume, volume: volume, isRecording: true)
    }
    
    static func text(text: String, duration: Int) -> VolcEngineSpeechContent {
        return VolcEngineSpeechContent(type: .text, text: text, duration: duration, isRecording: true)
    }
    
    static func recordStatus(isRecording: Bool) -> VolcEngineSpeechContent {
        return VolcEngineSpeechContent(type: .recordStatus, isRecording: isRecording)
    }
    
    enum CodingKeys: String, CodingKey {
        case type
        case volume
        case text
        case duration
        case isRecording
    }
}

File: android/src/main/kotlin/com/wood/volc_engine_asr/VolcEngineSpeechContent.kt
package com.wood.volc_engine_asr

import com.google.gson.Gson

enum class VolcEngineSpeechContentType {
    VOLUME, TEXT, RECORD_STATUS
}

private val gson = Gson()

data class VolcEngineSpeechContent(
    val type: VolcEngineSpeechContentType,
    val volume: Double = 0.0,
    val text: String = "",
    val duration: Int = 0,
    val isRecording: Boolean = true,
) {

    fun toJson(): String = gson.toJson(mapOf(
        "type" to type.ordinal,
        "volume" to volume,
        "text" to text,
        "duration" to duration,
        "isRecording" to isRecording
    ))

    companion object {
        fun volume(volume: Double): VolcEngineSpeechContent {
            return VolcEngineSpeechContent(VolcEngineSpeechContentType.VOLUME, volume)
        }

        fun text(text: String, duration: Int): VolcEngineSpeechContent {
            return VolcEngineSpeechContent(VolcEngineSpeechContentType.TEXT, text = text, duration = duration)
        }

        fun recordStatus(isRecording: Boolean): VolcEngineSpeechContent {
            return VolcEngineSpeechContent(VolcEngineSpeechContentType.RECORD_STATUS, isRecording = isRecording)
        }
    }
}
